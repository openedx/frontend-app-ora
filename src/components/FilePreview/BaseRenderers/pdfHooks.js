import { useRef } from 'react';

import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

export const stateKeys = StrictDict({
  pageNumber: 'pageNumber',
  numPages: 'numPages',
  relativeHeight: 'relativeHeight',
});

export const initialState = {
  pageNumber: 1,
  numPages: 1,
  relativeHeight: 1,
};

export const safeSetPageNumber = ({ numPages, rawSetPageNumber }) => (pageNumber) => {
  if (pageNumber > 0 && pageNumber <= numPages) {
    rawSetPageNumber(pageNumber);
  }
};

export const rendererHooks = ({
  onError,
  onSuccess,
}) => {
  const [pageNumber, rawSetPageNumber] = useKeyedState(stateKeys.pageNumber, initialState.pageNumber);
  const [numPages, setNumPages] = useKeyedState(stateKeys.numPages, initialState.numPages);
  const [relativeHeight, setRelativeHeight] = useKeyedState(
    stateKeys.relativeHeight,
    initialState.relativeHeight,
  );

  const setPageNumber = safeSetPageNumber({ numPages, rawSetPageNumber });

  const wrapperRef = useRef();

  return {
    pageNumber,
    numPages,
    relativeHeight,
    wrapperRef,
    onDocumentLoadSuccess: (args) => {
      onSuccess();
      setNumPages(args.numPages);
    },
    onLoadPageSuccess: (page) => {
      const pageWidth = page.view[2];
      const pageHeight = page.view[3];
      const wrapperHeight = wrapperRef.current.getBoundingClientRect().width;
      const newHeight = (wrapperHeight * pageHeight) / pageWidth;
      setRelativeHeight(newHeight);
    },
    onDocumentLoadError: onError,
    onInputPageChange: ({ target: { value } }) => setPageNumber(parseInt(value, 10)),
    onPrevPageButtonClick: () => setPageNumber(pageNumber - 1),
    onNextPageButtonClick: () => setPageNumber(pageNumber + 1),
    hasNext: pageNumber < numPages,
    hasPrev: pageNumber > 1,
  };
};
