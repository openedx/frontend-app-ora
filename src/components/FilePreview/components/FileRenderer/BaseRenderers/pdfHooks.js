import { useRef, useState } from 'react';

export const initialState = {
  pageNumber: 1,
  numPages: 1,
  relativeHeight: 1,
};

export const safeSetPageNumber = ({ numPages, rawSetPageNumber }) => (pageNumber) => {
  if (pageNumber > 0 && pageNumber <= numPages) {
    rawSetPageNumber(pageNumber);
  }
};

export const usePDFRendererData = ({
  onError,
  onSuccess,
}) => {
  const [pageNumber, rawSetPageNumber] = useState(initialState.pageNumber);
  const [numPages, setNumPages] = useState(initialState.numPages);
  const [relativeHeight, setRelativeHeight] = useState(initialState.relativeHeight);

  const setPageNumber = safeSetPageNumber({ numPages, rawSetPageNumber });

  const wrapperRef = useRef();

  return {
    pageNumber,
    numPages,
    relativeHeight,
    wrapperRef,
    onDocumentLoadSuccess: (args) => {
      onSuccess();
      setNumPages(args.numPages);
    },
    onLoadPageSuccess: (page) => {
      const pageWidth = page.view[2];
      const pageHeight = page.view[3];
      const wrapperHeight = wrapperRef.current.getBoundingClientRect().width;
      const newHeight = (wrapperHeight * pageHeight) / pageWidth;
      setRelativeHeight(newHeight);
    },
    onDocumentLoadError: onError,
    onInputPageChange: ({ target: { value } }) => setPageNumber(parseInt(value, 10)),
    onPrevPageButtonClick: () => setPageNumber(pageNumber - 1),
    onNextPageButtonClick: () => setPageNumber(pageNumber + 1),
    hasNext: pageNumber < numPages,
    hasPrev: pageNumber > 1,
  };
};
