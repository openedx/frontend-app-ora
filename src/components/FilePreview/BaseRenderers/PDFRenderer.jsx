import React from 'react';
import PropTypes from 'prop-types';

import { pdfjs, Document, Page } from 'react-pdf';
import {
  Icon, Form, ActionRow, IconButton,
} from '@edx/paragon';
import { ChevronLeft, ChevronRight } from '@edx/paragon/icons';
import pdfjsWorker from 'pdfjs-dist/build/pdf.worker.min';

import 'react-pdf/dist/esm/Page/TextLayer.css';
import 'react-pdf/dist/esm/Page/AnnotationLayer.css';

import { rendererHooks } from './pdfHooks';

pdfjs.GlobalWorkerOptions.workerSrc = pdfjsWorker;

/**
 * <PDFRenderer />
 */
export const PDFRenderer = ({
  onError,
  onSuccess,
  url,
}) => {
  const {
    pageNumber,
    numPages,
    relativeHeight,
    wrapperRef,
    onDocumentLoadSuccess,
    onLoadPageSuccess,
    onDocumentLoadError,
    onInputPageChange,
    onNextPageButtonClick,
    onPrevPageButtonClick,
    hasNext,
    hasPrev,
  } = rendererHooks({ onError, onSuccess });

  return (
    <div ref={wrapperRef} className="pdf-renderer">
      <Document
        file={url}
        onLoadSuccess={onDocumentLoadSuccess}
        onLoadError={onDocumentLoadError}
      >
        {/* <Outline /> */}
        <div className="page-wrapper" style={{ height: relativeHeight }}>
          <Page height={relativeHeight} pageNumber={pageNumber} onLoadSuccess={onLoadPageSuccess} />
        </div>
      </Document>
      <ActionRow className="d-flex justify-content-center m-0">
        <IconButton
          size="inline"
          alt="previous pdf page"
          iconAs={Icon}
          src={ChevronLeft}
          disabled={!hasPrev}
          onClick={onPrevPageButtonClick}
        />
        <Form.Group className="d-flex align-items-center m-0">
          <Form.Label isInline>Page </Form.Label>
          <Form.Control
            type="number"
            min={0}
            max={numPages}
            value={pageNumber}
            onChange={onInputPageChange}
          />
          <Form.Label isInline> of {numPages}</Form.Label>
        </Form.Group>
        <IconButton
          size="inline"
          alt="next pdf page"
          iconAs={Icon}
          src={ChevronRight}
          disabled={!hasNext}
          onClick={onNextPageButtonClick}
        />
      </ActionRow>
    </div>
  );
};

PDFRenderer.defaultProps = {};

PDFRenderer.propTypes = {
  url: PropTypes.string.isRequired,
  onError: PropTypes.func.isRequired,
  onSuccess: PropTypes.func.isRequired,
};

export default PDFRenderer;
