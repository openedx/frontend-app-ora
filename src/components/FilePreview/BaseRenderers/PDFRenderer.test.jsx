import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import PDFRenderer from './PDFRenderer';

import * as hooks from './pdfHooks';

jest.mock('react-pdf', () => ({
  pdfjs: { GlobalWorkerOptions: {} },
  Document: () => 'Document',
  Page: () => 'Page',
}));

jest.mock('./pdfHooks', () => ({
  rendererHooks: jest.fn(),
}));

describe('PDF Renderer Component', () => {
  const props = {
    url: 'some_url.pdf',
    onError: jest.fn().mockName('this.props.onError'),
    onSuccess: jest.fn().mockName('this.props.onSuccess'),
  };
  const hookProps = {
    pageNumber: 1,
    numPages: 10,
    relativeHeight: 200,
    wrapperRef: { current: 'hooks.wrapperRef' },
    onDocumentLoadSuccess: jest.fn().mockName('hooks.onDocumentLoadSuccess'),
    onLoadPageSuccess: jest.fn().mockName('hooks.onLoadPageSuccess'),
    onDocumentLoadError: jest.fn().mockName('hooks.onDocumentLoadError'),
    onInputPageChange: jest.fn().mockName('hooks.onInputPageChange'),
    onNextPageButtonClick: jest.fn().mockName('hooks.onNextPageButtonClick'),
    onPrevPageButtonClick: jest.fn().mockName('hooks.onPrevPageButtonClick'),
    hasNext: true,
    hasPref: false,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('snapshots', () => {
    test('first page, prev is disabled', () => {
      hooks.rendererHooks.mockReturnValue(hookProps);
      expect(shallow(<PDFRenderer {...props} />).snapshot).toMatchSnapshot();
    });
    test('on last page, next is disabled', () => {
      hooks.rendererHooks.mockReturnValue({
        ...hookProps,
        pageNumber: hookProps.numPages,
        hasNext: false,
        hasPrev: true,
      });
      expect(shallow(<PDFRenderer {...props} />).snapshot).toMatchSnapshot();
    });
  });
});
