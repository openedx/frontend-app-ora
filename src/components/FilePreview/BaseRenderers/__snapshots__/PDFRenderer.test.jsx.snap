// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDF Renderer Component snapshots first page, prev is disabled 1`] = `
<div
  className="pdf-renderer"
>
  <Document
    file="some_url.pdf"
    onLoadError={[MockFunction hooks.onDocumentLoadError]}
    onLoadSuccess={[MockFunction hooks.onDocumentLoadSuccess]}
  >
    <div
      className="page-wrapper"
      style={
        Object {
          "height": 200,
        }
      }
    >
      <Page
        height={200}
        onLoadSuccess={[MockFunction hooks.onLoadPageSuccess]}
        pageNumber={1}
      />
    </div>
  </Document>
  <ActionRow
    className="d-flex justify-content-center m-0"
  >
    <IconButton
      alt="previous pdf page"
      disabled={true}
      iconAs="Icon"
      onClick={[MockFunction hooks.onPrevPageButtonClick]}
      size="inline"
      src={[Function]}
    />
    <Form.Group
      className="d-flex align-items-center m-0"
    >
      <Form.Label
        isInline={true}
      >
        Page 
      </Form.Label>
      <Form.Control
        max={10}
        min={0}
        onChange={[MockFunction hooks.onInputPageChange]}
        type="number"
        value={1}
      />
      <Form.Label
        isInline={true}
      >
         of 
        10
      </Form.Label>
    </Form.Group>
    <IconButton
      alt="next pdf page"
      disabled={false}
      iconAs="Icon"
      onClick={[MockFunction hooks.onNextPageButtonClick]}
      size="inline"
      src={[Function]}
    />
  </ActionRow>
</div>
`;

exports[`PDF Renderer Component snapshots on last page, next is disabled 1`] = `
<div
  className="pdf-renderer"
>
  <Document
    file="some_url.pdf"
    onLoadError={[MockFunction hooks.onDocumentLoadError]}
    onLoadSuccess={[MockFunction hooks.onDocumentLoadSuccess]}
  >
    <div
      className="page-wrapper"
      style={
        Object {
          "height": 200,
        }
      }
    >
      <Page
        height={200}
        onLoadSuccess={[MockFunction hooks.onLoadPageSuccess]}
        pageNumber={10}
      />
    </div>
  </Document>
  <ActionRow
    className="d-flex justify-content-center m-0"
  >
    <IconButton
      alt="previous pdf page"
      disabled={false}
      iconAs="Icon"
      onClick={[MockFunction hooks.onPrevPageButtonClick]}
      size="inline"
      src={[Function]}
    />
    <Form.Group
      className="d-flex align-items-center m-0"
    >
      <Form.Label
        isInline={true}
      >
        Page 
      </Form.Label>
      <Form.Control
        max={10}
        min={0}
        onChange={[MockFunction hooks.onInputPageChange]}
        type="number"
        value={10}
      />
      <Form.Label
        isInline={true}
      >
         of 
        10
      </Form.Label>
    </Form.Group>
    <IconButton
      alt="next pdf page"
      disabled={true}
      iconAs="Icon"
      onClick={[MockFunction hooks.onNextPageButtonClick]}
      size="inline"
      src={[Function]}
    />
  </ActionRow>
</div>
`;
